pkg/operator/controller/extension/virtualcluster/reconciler.go

```diff
@@ -0,0 +1,324 @@
+	ConditionReconcileSuccess = "ReconcileSuccessful"
+	// ConditionDeleteSuccessful is the condition type for when the virtual cluster resources successfully delete.
+	ConditionDeleteSuccessful = "DeleteSuccessful"
+	// requeueGardenResourceNotReady is the time after which an extension will be requeued, if the Garden resource was not ready during its reconciliation.
+	requeueGardenResourceNotReady = 10 * time.Second
+)
+
+// Reconciler reconciles Extensions.
+type Reconciler struct {
+	RuntimeClient   client.Client
+	Config          config.OperatorConfiguration
+	Clock           clock.Clock
+	Recorder        record.EventRecorder
+	GardenNamespace string
+	// GardenClientMap is the ClientMap used to communicate with the virtual garden cluster. It should be set by AddToManager function but the field is still public for use in tests.
+	GardenClientMap clientmap.ClientMap
+}
+
+// Reconcile performs the main reconciliation logic.
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := logf.FromContext(ctx)
+
+	ctx, cancel := controllerutils.GetMainReconciliationContext(ctx, controllerutils.DefaultReconciliationTimeout)
+	defer cancel()
+
+	extension := &operatorv1alpha1.Extension{}
+	if err := r.RuntimeClient.Get(ctx, request.NamespacedName, extension); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.V(1).Info("Object is gone, stop reconciling")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("error retrieving object from store: %w", err)
+	}
+
+	gardenList := &operatorv1alpha1.GardenList{}
+	// We limit one result because we expect only a single Garden object to be there.
+	if err := r.RuntimeClient.List(ctx, gardenList, client.Limit(1)); err != nil {
+		return reconcile.Result{}, fmt.Errorf("error retrieving Garden object: %w", err)
+	}
+	if len(gardenList.Items) == 0 {
+		// in case a garden resource does not exist or is deleted, update the condition, remove the finalizers and exit early.
+		log.Info("No Garden found")
+		conditions := NewVirtualClusterConditions(r.Clock, extension.Status)
+		conditions.virtualClusterReconciled = v1beta1helper.UpdatedConditionWithClock(r.Clock, conditions.virtualClusterReconciled, gardencorev1beta1.ConditionFalse, ConditionNoGardenFound, "no garden found")
+		if err := r.updateExtensionStatus(ctx, log, extension, conditions); err != nil {
+			log.Error(err, "Failed to update Extension status")
+		}
+		return reconcile.Result{}, r.removeFinalizer(ctx, log, extension)
+	}
+
+	garden := &gardenList.Items[0]
+	// check Garden's last operation status. If the last operation is a successful reconciliation, we can proceed to install the extensions to the virtual garden cluster.
+	switch lastOperation := garden.Status.LastOperation; {
+	case lastOperation == nil:
+		fallthrough
+	case lastOperation.Type == gardencorev1beta1.LastOperationTypeReconcile && lastOperation.State != gardencorev1beta1.LastOperationStateSucceeded:
```

Comment:
Interesting style 🙈 Personally, I was not aware of this `fallthrough` behaviour. While this works, I'd prefer to make it more explicit:

```suggestion
	case lastOperation == nil || (lastOperation.Type == gardencorev1beta1.LastOperationTypeReconcile && lastOperation.State != gardencorev1beta1.LastOperationStateSucceeded):
```

I understand that this might not necessarily ease readability, but it at least does not require to look up the `fallthrough` behaviour 😄 But maybe it's only me 😉